pkg/controller/cabundleinjector/admissionwebhook.go

```diff
@@ -0,0 +1,161 @@
+	return &validatingWebhookConfigAccessor{
+		ValidatingWebhookConfiguration: a.ValidatingWebhookConfiguration.DeepCopy(),
+	}
+}
+
+func (a *validatingWebhookConfigAccessor) GetObject() *admissionregv1.ValidatingWebhookConfiguration {
+	return a.ValidatingWebhookConfiguration
+}
+
+type cachedWebhookConfigGetter[T admissionregv1.MutatingWebhookConfiguration | admissionregv1.ValidatingWebhookConfiguration] interface {
+	Get(name string) (*T, error)
+}
+
+type webhookConfigUpdater[T admissionregv1.MutatingWebhookConfiguration | admissionregv1.ValidatingWebhookConfiguration] interface {
+	Update(ctx context.Context, webhookConfig *T, updateOptions metav1.UpdateOptions) (*T, error)
+}
+
+// webhookCABundleInjector creates a controller that injects the service-ca bundle
+// to validating and mutating webhookconfigurations
+type webhookCABundleInjector[T admissionregv1.MutatingWebhookConfiguration | admissionregv1.ValidatingWebhookConfiguration] struct {
+	webhookConfigType         string
+	client                    webhookConfigUpdater[T]
+	lister                    cachedWebhookConfigGetter[T]
+	webhookConfigAccessorInit func(*T) webhookConfigAccessor[T]
+	caBundle                  []byte
+}
+
+func (bi *webhookCABundleInjector[T]) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	webhookConfig, err := bi.lister.Get(syncCtx.QueueKey())
+	if apierrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	webhookConfigAccessor := bi.webhookConfigAccessorInit(webhookConfig)
+
+	webhooksNeedingUpdate := []int{}
+	for i := 0; i < webhookConfigAccessor.WebhooksLen(); i++ {
+		webhookClientConfig := webhookConfigAccessor.GetWebhookClientCA(i)
+		if !bytes.Equal(webhookClientConfig.CABundle, bi.caBundle) {
+			webhooksNeedingUpdate = append(webhooksNeedingUpdate, i)
+		}
+	}
+	if len(webhooksNeedingUpdate) == 0 {
+		return nil
+	}
+
+	klog.Infof("updating %s %s with the service signing CA bundle", bi.webhookConfigType, webhookConfigAccessor.GetName())
+
+	// make a copy to avoid mutating cache state
+	webhookConfigCopy := webhookConfigAccessor.DeepCopy()
+	for _, i := range webhooksNeedingUpdate {
+		webhookConfigCopy.GetWebhookClientCA(i).CABundle = bi.caBundle
+	}
+	_, err = bi.client.Update(ctx, webhookConfigCopy.GetObject(), metav1.UpdateOptions{})
+	return err
+}
+
+func newMutatingWebhookInjectorConfig(config *caBundleInjectorConfig) controllerConfig {
+	informer := config.kubeInformers.Admissionregistration().V1().MutatingWebhookConfigurations()
+	syncer := &webhookCABundleInjector[admissionregv1.MutatingWebhookConfiguration]{
+		webhookConfigType:         "mutatingwebhookconfiguration",
+		client:                    config.kubeClient.AdmissionregistrationV1().MutatingWebhookConfigurations(),
+		lister:                    informer.Lister(),
+		webhookConfigAccessorInit: newMutatingWebhookAccessor,
```

Comment:
let's call this `newWebhookAccessor`, this naming provides a nice abstraction to the concrete implementations `newMutatingWebhookAccessor` and `newValidatingWebhookAccessor`.

Reply:
note: initialization is different from construction and usually comes after it.